src/envs/pybullet_cover.py

```diff
@@ -0,0 +1,313 @@
+            self._target_id_to_target[target_id] = target_obj
+            tx = self._workspace_x
+            # De-normalize target y to actual coordinates.
+            y_norm = state.get(target_obj, "pose")
+            ty = self._y_lb + (self._y_ub - self._y_lb) * y_norm
+            tz = self._table_height + self._obj_len_hgt * 0.5
+            p.resetBasePositionAndOrientation(
+                target_id, [tx, ty, tz], [0.0, 0.0, 0.0, 1.0],
+                physicsClientId=self._physics_client_id)
+
+        # Draw hand regions as debug lines.
+        p.removeAllUserDebugItems(physicsClientId=self._physics_client_id)
+        for hand_lb, hand_rb in self._get_hand_regions(state):
+            # De-normalize hand bounds to actual coordinates.
+            y_lb = self._y_lb + (self._y_ub - self._y_lb) * hand_lb
+            y_rb = self._y_lb + (self._y_ub - self._y_lb) * hand_rb
+            p.addUserDebugLine(
+                [self._workspace_x, y_lb, self._table_height + 1e-4],
+                [self._workspace_x, y_rb, self._table_height + 1e-4],
+                [0.0, 0.0, 1.0],
+                lineWidth=5.0,
+                physicsClientId=self._physics_client_id)
+
+    def step(self, action: Action) -> State:
+        # In the cover environment, we need to first check the hand region
+        # constraint before we can call PyBullet.
+        # Use self._pybullet_robot2 to run forward kinematics, since that
+        # method shouldn't be run on the client that is doing simulation.
+        _, ry, rz = self._pybullet_robot2.forward_kinematics(action.arr)
+        hand = (ry - self._y_lb) / (self._y_ub - self._y_lb)
+        hand_regions = self._get_hand_regions(self._current_state)
+        # If we're going down to grasp, we need to be in a hand region.
+        # Otherwise, we don't care if we're between hand regions.
+        z_thresh = (self._pickplace_z + self._workspace_z) / 2
+        if rz < z_thresh and not any(hand_lb <= hand <= hand_rb
+                                     for hand_lb, hand_rb in hand_regions):
+            # The constraint is violated, so no-op.
+            return self._current_state.copy()
+        return super().step(action)
+
+    def _get_state(self) -> State:
+        state_dict = {}
+        max_width = max(max(CFG.cover_block_widths),
+                        max(CFG.cover_target_widths))
+
+        # Get robot state.
+        rx, ry, rz, _ = self._pybullet_robot.get_state()
+        hand = (ry - self._y_lb) / (self._y_ub - self._y_lb)
+        state_dict[self._robot] = np.array([hand, rx, rz], dtype=np.float32)
+        joint_state = self._pybullet_robot.get_joints()
+
+        # Get block states.
+        for block_id, block in self._block_id_to_block.items():
+            width_unnorm = p.getVisualShapeData(
+                block_id, physicsClientId=self._physics_client_id)[0][3][1]
+            width = width_unnorm / self._max_obj_width * max_width
```

Comment:
why can't we just use `CFG.cover_block_widths` directly to get the block widths?